### PR TITLE
Fast decimal division

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 1]
```diff
@@ -427,9 +427,9 @@ public static void multiply(Slice left, Slice right, Slice result)
     public static int compare(Slice left, Slice right)
     {
         boolean leftStrictlyNegative = isStrictlyNegative(left);
-        boolean rightNegative = isNegative(right);
+        boolean rightStrictlyNegative = isStrictlyNegative(right);
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:489]: maybe `s/isStrictlyNegative/isNonZeroNegative/`

Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:489]: add unit tests

**Comment by kokosing** [comment]
looks good

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 1 above)
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:489]: +1 unit tests

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 2]
```diff
@@ -680,6 +682,11 @@ private static boolean divideCheckRound(Slice decimal, int divisor, Slice result
     // visible for testing
     static int divide(Slice decimal, int divisor, Slice result)
     {
+        if (divisor == 0) {
```
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:1112]: unit tests?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 3]
```diff
@@ -380,6 +380,10 @@ public static Slice multiply(Slice left, Slice right)
 
     public static void multiply(Slice left, Slice right, Slice result)
     {
+        boolean fourIntsResultExpected = result.length() == NUMBER_OF_LONGS * Long.BYTES;
+        boolean eightIntsResultExpected = result.length() >= NUMBER_OF_LONGS * Long.BYTES * 2;
```
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:387]: ==?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 4]
```diff
@@ -380,6 +380,10 @@ public static Slice multiply(Slice left, Slice right)
...
+        checkArgument(fourIntsResultExpected || eightIntsResultExpected);
```
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:388]: xor?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 5]
```diff
@@ -730,6 +730,117 @@ static void shiftRight(Slice decimal, int rightShifts, boolean roundUp, Slice re
         pack(result, lo, hi, negative);
     }
 
+    public static Slice divideRoundUp(long dividend, int dividendScaleFactor, long divisor)
```
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:733]: I would put those wrapper methods in `DecimalOperators.java` as it is done with other operators. That would leave only the important function `private static Slice divideRoundUp(long dividendLow, long dividendHi, int dividendScaleFactor, long divisorLow, long divisorHi)` here.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 6]
```diff
@@ -730,6 +730,117 @@ static void shiftRight(Slice decimal, int rightShifts, boolean roundUp, Slice re
...
+    }
+
+    public static Slice divideRoundUp(Slice dividend, int dividendScaleFactor, Slice divisor)
+    {
+        return divideRoundUp(
+                getRawLong(dividend, 0), getRawLong(dividend, 1), dividendScaleFactor,
+                getRawLong(divisor, 0), getRawLong(divisor, 1));
+    }
+
+    private static Slice divideRoundUp(long dividendLow, long dividendHi, int dividendScaleFactor, long divisorLow, long divisorHi)
```
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java:761]: `s/Hi/High/g`
`s/hi/high/g`

Karol has some pull request that fixes that in rest of the code, so don't introduce new code with abbreviations. Fix this in all new code.

**Comment by pnowojski** [comment]
please, provide benchmark results

**Comment by arhimondr** [comment]
Was merged to `feature_fast_decimal_v3`

**Pull request closed without merging**